torch/_inductor/graph.py

```diff
@@ -856,6 +860,21 @@ def placeholder(self, target: str, args, kwargs):
         self.graph_inputs[target] = tensor
         self.graph_inputs_original[target] = tensor.data.data
         self.add_device_info(example.device)
+
+        # Note: [Input Alignment handling in Inductor]
+        # Alignment matters for generating efficient code. Some operations,
+        # e.g. vectorized loads, can only be performed on aligned inputs.
+        #
+        # But if we codegen assuming aligned inputs and then get unaligned
+        # inputs at runtime, then we are forced to clone - which is bad for
+        # both perf and memory usage.
+        #
+        # One option would be to guard on storage_offset%ALIGNMENT, and then
+        # codegen based on this. But storage_offset guards turned out to be
+        # expensive and cause recompiles; Instead, we're generating code
+        # based on the alignment of the example input without guarding.
+        if should_assume_input_aligned(example):
+            self.aligned_inputs.add(target)
```

Comment:
```
import torch

@torch.compile()
def foo(x):
    return x.add_(2)

y = torch.zeros([4], device="cuda")
z = y[1:]
foo(z)
print(z)
```

Let's also not assume alignment on mutated inputs to fix the above ^^ issue

Reply:
what's the issue here? I ran this and it seems to return `[2., 2., 2.]` as expected

(also - didn't realize inductor had to deal with mutations? I thought aot autograd / functionalization would deal with this?)

Reply:
Also for this, I think we'd need to propagate this info to compile_fx.py, which isn't being done right now. I think this is somewhat non-trivial:

- right now, graph.py aligned input handling is only used to do codegen
- separately, in compile_fx.py we also check should_assume_input_aligned to see whether we should do a clone
- it seems like if we want to rely on graph.py artifacts in compile_fx.py, this is a bit difficult because we cache the resulting artifacts sometimes (and in the cached case I don't think we have access to self.aligned_inputs in compile_fx.py?)

Reply:
Try running this :
import torch

@torch.compile()
def foo(x):
    return x.add_(2)

y = torch.zeros([16], device="cuda")
# specialize on aligned
foo(y[8:12])

# re-run with unaligned will clone input
z = y[1:5]
print(z)
```

Yes - for inference (and maybe training? not sure) we now add back in the copy_ kernels into the graph instead of running them in aot autograd

Reply:
Hmm maybe we can do this in a separate PR. not sure if this will lead to perf decrease.

Reply:
> we now add back in the copy_ kernels into the graph instead of running them in aot autograd

Yup we now do this for training too (the tldr is that as long as the input mutation doesn't mutate the grad_fn of the input, we try to keep it in the graph)